### ISSUE TITLE
Performance degradation

### ISSUE THREAD

**Issue by jpvandervelden** [dmlc/xgboost#7138]
Hi,

We're using XGBoost in our Python project to train our models. We've been using version 1.2.1 for some time. It turned out there was security issue with that version. We upgraded to 1.4.1 but soon found out that the performance degraded from 9 mins to 57 mins. I did some tests to see what the duration was for the different versions.

<html>
<body>

Version |  Duration | Model Score | Safe
-- |  -- | -- | --
1.2.1 | 9 mins | 0.8008968609865471 | No
1.3.1 | 57 mins | 0.8026905829596414 | Yes
1.3.3 | 44 mins | 0.8026905829596414 | Yes
1.4.1 | 57 mins | 0.8026905829596414 | Yes
1.4.2 | 56 mins | 0.8026905829596414 | Yes

</body>
</html>

I couldn't find any reports about other people experiencing this.

I hope there's some way to overcome this...

Cheers,

JP

**Comment by trivialfis** [comment]
Could you please share the training parameter and the data shape along with its sparsity?

**Comment by trivialfis** [comment]
Also, which interface are you using?

**Comment by jpvandervelden** [comment]
Here some snippets from the code:

`from xgboost import XGBClassifier

classifier = XGBClassifier(use_label_encoder=False, eval_metric = 'merror')

parameters_xgb_computer = {
              "class__regressor__n_estimators": [10, 20, 30], #number of estimaters for GradientBoostingClassifier
              "class__regressor__learning_rate": [0.1, 0.2, 0.3], #learning rate for GradientBoostingClassifier
              "class__regressor__max_depth": [2, 3, 9]
              }`

The shape is:  (51345,, 9452)

It's a sparse matrix as TF-IDF and OneHotEncoding are used. I'll try to come up with the sparsity measure.

**Comment by trivialfis** [comment]
Are you using scikit-learn grid search?

**Comment by jpvandervelden** [comment]
yep... from sklearn.model_selection import GridSearchCV

**Comment by trivialfis** [comment]
Could you plese set `n_jobs` to 1 for xgb classifier?

**Comment by jpvandervelden** [comment]
I can. And then check the durations once more?

**Comment by trivialfis** [comment]
Yes.

**Comment by jpvandervelden** [comment]
Using hyperparameter n_jobs=1 the outcomes are:

<html>
<body>


Version |  Duration | Model Score | Safe
-- |  -- | -- | --
1.2.1 | 10 mins | 0.8008968609865471 | No
1.3.3 | <still running> |  | Yes
1.4.2 | 57 mins | 0.8026905829596414 | Yes


</body>
</html>

**Comment by trivialfis** [comment]
Thanks for the infomation.  I will try to reproduce and bisect the changes once you have the sparsity information.

**Comment by jpvandervelden** [comment]
SPARSITY: 0.9804190291345377

**Comment by trivialfis** [comment]
> SPARSITY: 0.9804190291345377

Hmm, only 2 percent of the elements.  Might be caused by changes in DMatrix.  Anyway, will bisect it.

**Comment by castorfou** [comment]
I have the same interest.
With 0.90 training takes about 2 mins. And in version 1.4.0 after 14 hours it is not completed.
Training with gridsearch from scikit-learn, large n_estimators are specially long:
[CV] END colsample_bytree=0.6, eta=0.01, max_depth=10, n_estimators=1500; total time=131.0min

**Comment by trivialfis** [comment]
@jpvandervelden @castorfou  How many classes do you have?  Are the labels encoded?

**Comment by castorfou** [comment]
I use a regressor here. (xgb.XGBRegressor)
And the scikit-learn way to train (not using DMatrix):

```python
    SEED=42
    xgb_model=xgb.XGBRegressor(gamma=0,
                     min_child_weight=1.5,
                     reg_alpha=0.75,
                     reg_lambda=0.45,
                     subsample=0.6,
                     seed=SEED,
                     objective='reg:squarederror',
                              verbosity=1)
    grid_xgb = GridSearchCV(estimator=xgb_model, param_grid=xgb_param_grid,
                            scoring='explained_variance', cv=10, verbose=1, n_jobs=-1)

#   Create the training and test sets
    X_train, X_test, y_train, y_test= train_test_split(X, y, test_size=0.2, random_state=SEED)
    if (feature_selection is not None):
        X_train_fs, X_test_fs, fs = feature_selection(X_train, y_train, X_test)
        X_train = X_train_fs
        X_test = X_test_fs


    grid_xgb.fit(X_train,y_train)
    model = grid_xgb.best_estimator_
    print('best score (evs):',grid_xgb.best_score_)
```

**Comment by trivialfis** [comment]
Thanks for sharing.  Looking into the difference.

**Comment by trivialfis** [comment]
@jpvandervelden  Likely that you are using the Exact tree method, could you please consider trying faster approximation methods like `hist` and `gpu_hist`?

**Comment by jpvandervelden** [comment]
If I understand correctly Exact is the vanilla XGBoosting tree algorithm. As I didn't specify anything I guess I'm using the Exact algorithm then. I will have a go with "hist" and the machine doesn't have a GPU.

**Comment by trivialfis** [comment]
We have documents on various tree methods: https://xgboost.readthedocs.io/en/latest/treemethod.html

**Comment by jpvandervelden** [comment]
Thanks. I looked into those and see the different options..   started the training.. will report the results as they come in..

**Comment by trivialfis** [comment]
It's a little bit unusual for this level of sparsity, so we might not be constantly looking at these kinds of datasets.  I will create some PRs and try to bring back some performance in the coming days.

**Comment by jpvandervelden** [comment]
#labels = 34

**Comment by trivialfis** [comment]
@ShvetsKS Could you please help take a look into the hist performance for sparse datasets?  On my machine the following example is slower than exact:

``` python
from scipy import sparse
import numpy as np
import xgboost as xgb
from joblib import Memory
from time import time


memory = Memory("./cachedir")

rng = np.random.RandomState(1994)
sparsity = 0.9804190291345377


@memory.cache
def make_blob():
    X = sparse.random(m=51345, n=9452, density=1.0 - sparsity, random_state=rng)
    y = rng.randint(0, high=10, size=X.shape[0])
    return X, y


def main():
    X, y = make_blob()
    X = X.tocsr()

    begin = time()
    clf = xgb.XGBClassifier(
        n_estimators=64, use_label_encoder=False, verbosity=1, n_jobs=16,
    )
    clf.fit(X, y, eval_metric='merror', eval_set=[(X, y)], verbose=True)
    print(clf.get_booster().save_config())
    end = time()

    print("Duration:", end - begin)


if __name__ == "__main__":
    main()
```

**Comment by jpvandervelden** [comment]
Using parameter "tree_method": "hist" 

The column Duration is the time in minutes is the total time including overhead of starting the instance. The column Training is the actual time to train the model in seconds. I will add those times for the earlier results. 

<html>
<body>

<div class="table-wrap">

Version | Duration | Training | Model Score
-- | -- | -- | --
1.2.1 | 9 mins | 430 sec | 0.8008968609865471
1.3.3 | 43 mins | 2469 sec | 0.8026905829596414
1.4.1 | 57 mins | 3271 sec | 0.8026905829596414
1.4.2 | 44 mins | 2479 sec | 0.8026905829596414


</div>
</body>
</html>

**Comment by ShvetsKS** [comment]
@trivialfis, this workload has too few samples for `hist` method and top hotspots are `Evaluate` and `Sync` histograms kernels (doesn't depends on number of samples).
For 200k samples training times are similar (and accuracy is better for `hist`). For `m=513450` `hist` method 1.7 times outperforms `exact` in this case.
The Evaluate and Sync histograms kernels can be easily optimized by merging these two kernels and adding dynamic scheduling (as not balanced work is asigned to each thread due to different number of bins for each feature). With prototyped optimization `hist` method outperforms `exact` on 150k samples already.

**Comment by trivialfis** [comment]
Yeah, these sparse datasets are like the URL dataset which is wide and sparse due to encoding.   Dynamic scheduling sounds like a good solution.

**Comment by ShvetsKS** [comment]
> Yeah, these sparse datasets are like the URL dataset which is wide and sparse due to encoding. Dynamic scheduling sounds like a good solution.

on the URL dataset 'hist' method outperforms `exact`, just checked it.

**Comment by trivialfis** [comment]
I bisected the performance change to 2fcc4f28 (#6037), my fault as usual ...  I managed to reduce some overhead caused by conditional branches, but to be honest I'm a little bit reluctant to spend too much time on optimizing the exact tree method.  So I'm hoping the hist tree method can just outperform exact in this case.

**Comment by trivialfis** [comment]
Closing as we won't focus too much on the exact tree method in terms of computation performance.
